### PR TITLE
rdkit 2020.09.4 (new formula)

### PR DESCRIPTION
**Pull request by YoshitakaMo**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/rdkit.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    else
+      puts "assuming GCC C++ compiler"
+      # g++ doesn"t accept/need the option: -stdlib=libc++
+    end
+
+    # Get Python location
+    python3 = Formula["python@3.9"].opt_bin/"python3"
+    python_include =
+      Utils.safe_popen_read(python3, "-c", "from distutils import sysconfig;print(sysconfig.get_python_inc(True))")
+           .chomp
```
Comment (carlocab) [inline Formula/rdkit.rb:39]: There's a much simpler/idiomatic way to find the right `include` directory. Have a look at the `openimageio` formula.

File: Formula/rdkit.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    python_executable = Utils.safe_popen_read(python3, "-c", "import sys;print(sys.executable)").chomp
```
Comment (carlocab) [inline Formula/rdkit.rb:40]:
```suggestion
    python_executable = python3.opt_bin/"python3"
```

File: Formula/rdkit.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    numpy_include = Utils.safe_popen_read(python3, "-c", "import numpy ; print(numpy.get_include())").chomp
```
Comment (carlocab) [inline Formula/rdkit.rb:41]: See comment above re `openimageio`.

File: Formula/rdkit.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      -DRDK_BUILD_YAEHMOP_SUPPORT=ON
+      -DRDK_BUILD_FREESASA_SUPPORT=ON
+      -DBoost_NO_BOOST_CMAKE=ON
+      -DPYTHON_INCLUDE_DIR=#{python_include}
+      -DPYTHON_EXECUTABLE=#{python_executable}
+      -DPYTHON_NUMPY_INCLUDE_PATH=#{numpy_include}
+    ]
+
+    system "cmake", ".", *args
+    system "make", "-j#{ENV.make_jobs}"
```
Comment (carlocab) [inline Formula/rdkit.rb:64]:
```suggestion
    system "make"
```
`brew` will specify the correct `-j` flag on its own.

File: Formula/rdkit.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    system "make", "install"
+
+    # Remove the ghost .cmake files which will cause a warning if we install them to "lib"
+    rm_f Dir["#{lib}/*.cmake"]
+
+    mv "Code/PgSQL/rdkit/rdkit.sql.in", "Code/PgSQL/rdkit/rdkit--3.4.sql"
```
Comment (carlocab) [inline Formula/rdkit.rb:70]: This seems fragile. Where does the `3.4` come from?

File: Formula/rdkit.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    (share + "postgresql/extension").install "Code/PgSQL/rdkit/rdkit--3.4.sql"
+    (share + "postgresql/extension").install "Code/PgSQL/rdkit/rdkit.control"
+    (lib + "postgresql").install "Code/PgSQL/rdkit/rdkit.so"
```
Comment (carlocab) [inline Formula/rdkit.rb:73]:
```suggestion
    (share/"postgresql/extension").install "Code/PgSQL/rdkit/rdkit--3.4.sql"
    (share/"postgresql/extension").install "Code/PgSQL/rdkit/rdkit.control"
    (lib/"postgresql").install "Code/PgSQL/rdkit/rdkit.so"
```

**Review by carlocab — COMMENTED** [review]
Thanks for your PR, @YoshitakaMo.

File: Formula/rdkit.rb [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  end
+
+  def caveats
+    <<~EOS
+      You may need to add RDBASE to your environment variables.
+      For Bash, put something like this in your $HOME/.bashrc:
+        export RDBASE=#{HOMEBREW_PREFIX}/share/RDKit
+    EOS
+  end
```
[the remark this answers is not in the document]
Comment (YoshitakaMo) [reply Formula/rdkit.rb:72]: I think this `caveats` block is required for the users like `yubikey-agent` or `nodebrew` formulae.

File: Formula/rdkit.rb [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    ENV["CFLAGS"] = "-Wno-parentheses -Wno-logical-op-parentheses -Wno-format"
+    ENV["CXXFLAGS"] = "-std=c++11 -Wno-parentheses -Wno-logical-op-parentheses -Wno-format"
+    cpp_version_str = `c++ --version`
+    if cpp_version_str.include? "clang"
+      puts "detected clang C++ compiler"
+      ENV["CXXFLAGS"] += " -stdlib=libc++"
+    else
+      puts "assuming GCC C++ compiler"
+      # g++ doesn"t accept/need the option: -stdlib=libc++
+    end
```
[the remark this answers is not in the document]
Comment (YoshitakaMo) [reply Formula/rdkit.rb]: Thanks, @carlocab. However, I could not run this formula with this modification due to a syntax error.
So I modified explicitly as below:
```
  def install
    ENV.cxx11
    ENV.libcxx
    ENV.append "CFLAGS", "-Wno-parentheses -Wno-logical-op-parentheses -Wno-format"
    ENV.append "CXXFLAGS", "-Wno-parentheses -Wno-logical-op-parentheses -Wno-format"
```

File: Formula/rdkit.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (SMillerDev) [reply Formula/rdkit.rb:72]: Since the location is homebrew specific this is allowed @carlocab

File: Formula/rdkit.rb [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    system "make", "install"
+
+    # Remove the ghost .cmake files which will cause a warning if we install them to "lib"
+    rm_f Dir["#{lib}/*.cmake"]
+
+    # '3.8' is the default_version in Code/PgSQL/rdkit/rdkit.control
+    # This is a workaround to reproduce their filenames after installation.
+    (share/"postgresql/extension").install "Code/PgSQL/rdkit/rdkit--3.8.sql"
+    (share/"postgresql/extension").install "Code/PgSQL/rdkit/rdkit--3.8.control"
+    (lib/"postgresql").install "Code/PgSQL/rdkit/rdkit.so"
```
Comment (SMillerDev) [inline Formula/rdkit.rb:63]: This needs to be submitted and accepted as a patch upstream. We don't want manual patching in new formulae.

File: Formula/rdkit.rb (hunk 5 above)
Comment (YoshitakaMo) [reply Formula/rdkit.rb]: `3.4` represents the default_version in [Code/PgSQL/rdkit/rdkit.control](https://github.com/rdkit/rdkit/blob/master/Code/PgSQL/rdkit/rdkit.control). In this latest version, it should be `3.8`.
This is a workaround to reproduce their filenames after installation. But the truth is, I don't understand the need for it either.

File: Formula/rdkit.rb (hunk 9 above)
Comment (YoshitakaMo) [reply Formula/rdkit.rb:63]: OK, I will remove them and retry this PR because this workaround may no longer be required.

**Comment by YoshitakaMo** [comment]
In the [first PR](https://github.com/Homebrew/homebrew-core/runs/1949020835?check_suite_focus=true), there is an import error when `brew test` only on Mac 11-arm.

```
==> brew test --verbose rdkit
==> FAILED
==> Testing rdkit
/usr/bin/sandbox-exec -f /private/tmp/homebrew20210221-58613-y4zkmk.sb ruby -W1 -I $LOAD_PATH -- /opt/homebrew/Library/Homebrew/test.rb /opt/homebrew/Library/Taps/homebrew/homebrew-core/Formula/rdkit.rb --verbose
==> /opt/homebrew/opt/python@3.9/bin/python3 test.py 2>&1
Traceback (most recent call last):
  File "/private/tmp/rdkit-test-20210221-58614-189fhuk/test.py", line 1, in <module>
    from rdkit import Chem ; print(Chem.MolToSmiles(Chem.MolFromSmiles('C1=CC=CN=C1')))
  File "/opt/homebrew/lib/python3.9/site-packages/rdkit/__init__.py", line 2, in <module>
    from .rdBase import rdkitVersion as __version__
ImportError: dlopen(/opt/homebrew/lib/python3.9/site-packages/rdkit/rdBase.so, 2): Library not loaded: @rpath/libRDKitRDBoost.1.dylib
  Referenced from: /opt/homebrew/lib/python3.9/site-packages/rdkit/rdBase.so
  Reason: image not found
Error: rdkit: failed
An exception occurred within a child process:
  Test::Unit::AssertionFailedError: <0> expected but was
```

Curiously, the other tests finished successfully. Could someone help me?

File: Formula/rdkit.rb [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+    # Get Python location
+    python_executable = Formula["python@3.9"].opt_bin/"python3"
+    py3ver = Language::Python.major_minor_version Formula["python@3.9"].opt_bin/"python3"
+    py3prefix = Formula["python@3.9"].opt_frameworks/"Python.framework/Versions/#{py3ver}"
+    py3include = "#{py3prefix}/include/python#{py3ver}"
+
+    # set -DMAEPARSER and COORDGEN_FORCE_BUILD=ON to avoid conflicts with some formulae i.e. open-babel
+    args = std_cmake_args + %W[
+      -DRDK_INSTALL_INTREE=OFF
```
Comment (carlocab) [inline Formula/rdkit.rb:34]: Try adding `-DCMAKE_INSTALL_RPATH=#{opt_lib}` here.

**Comment by YoshitakaMo** [comment]
I couldn't address this issue. Could someone help me?
```
==> Starting build of rdkit
==> brew fetch --retry rdkit --build-bottle --force
==> brew install --only-dependencies --verbose --build-bottle rdkit
==> brew install --verbose --build-bottle rdkit
==> brew audit rdkit --online --new-formula
==> brew bottle --verbose --json rdkit
==> brew bottle --merge --write --no-commit ==> Determining rdkit bottle rebuild...
Bad address
==> FAILED
Error: No such file or directory @ rb_sysopen - ==> Determining rdkit bottle rebuild...
Bad address

==> brew uninstall --force rdkit
==> brew uninstall --force boost boost-python3 cairo cmake cython eigen fontconfig freetype gcc gdbm gettext glib gmp icu4c isl krb5 libffi libmpc libpng lzo meson mpfr ninja numpy openblas openssl@1.1 pcre pixman pkg-config postgresql@12 py3cairo python@3.9 readline sphinx-doc sqlite swig tcl-tk xz
Error: No such file or directory @ rb_sysopen - ==> Determining rdkit bottle rebuild...
Bad address
Error: Process completed with exit code 1.
1s
Run cat bottles/steps_output.txt
cat: bottles/steps_output.txt: No such file or directory
Error: Process completed with exit code 1.
```

File: Formula/rdkit.rb (hunk 10 above)
Comment (YoshitakaMo) [reply Formula/rdkit.rb:35]: Thank you, @carlocab. The issue was solved!

File: Formula/rdkit.rb [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+  head "https://github.com/rdkit/rdkit.git"
+
+  depends_on "cmake" => :build
+  depends_on "swig" => :build
+  depends_on "boost"
+  depends_on "boost-python3"
+  depends_on "eigen"
+  depends_on "freetype"
+  depends_on "numpy"
+  depends_on "postgresql@12"
```
Comment (SMillerDev) [inline Formula/rdkit.rb:16]: Does it not work with the latest?

Comment (YoshitakaMo) [reply Formula/rdkit.rb]: Thanks, @SMillerDev. This rdkit version was already compatible with postgresql 13.

**Review by carlocab — APPROVED** [review]
Thanks, @YoshitakaMo.

**Pull request closed without merging**